File: src/com/easy/HouseRobber_198.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.1
 */
public class HouseRobber_198 {

    // DP[i] = max(DP[i-2] + nums[i], DP[i-1])
    // Time Complexity: O(n)
    // Space complexity: O(1)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 33.5 MB, less than 100.00%
    public int rob(int[] nums) {
        int mRob = 0;
        int mNotRob = 0;
        for (int i = 0; i < nums.length; i++) {
            int mCurrent = Math.max(mNotRob + nums[i], mRob);
            mNotRob = mRob;
            mRob = mCurrent;
        }

        return mRob;
    }

    // Judge even & odd position
    // Time Complexity: O(n)
    // Space complexity: O(1)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 33.2 MB, less than 100.00%
    public int rob2(int[] nums) {
        int mEvenRob = 0, mOddRob = 0;
        for (int i = 0; i < nums.length; i++) {
            if ((i % 2) == 0) {
                mEvenRob = Math.max(mEvenRob + nums[i], mOddRob);
            } else {
                mOddRob = Math.max(mOddRob + nums[i], mEvenRob);
            }
        }

        return Math.max(mEvenRob, mOddRob);
    }

    // Use recursive with memorize
    // Time Complexity: O(n)
    // Space complexity: O(n)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 34.3 MB, less than 100.00%
    public int rob3(int[] nums) {
        Integer[] mMem = new Integer[nums.length];
        return rob(nums, nums.length - 1, mMem);
    }

    private int rob(int[] nums, int mIndex, Integer[] mMem) {
        if (mIndex < 0) {
            return 0;
        }
        if (mMem[mIndex] != null) {
            return mMem[mIndex];
        }

        mMem[mIndex] = Math.max(rob(nums, mIndex - 2, mMem) + nums[mIndex], rob(nums, mIndex - 1, mMem));
        return mMem[mIndex];
    }
}

File: Golang_src/easy/houseRobber_198.go
package easy

// Use DP:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 23.78%
func rob(nums []int) int {
	rob, notRob := 0, 0

	for _, num := range nums {
		cur := max(notRob+num, rob)
		notRob = rob
		rob = cur
	}

	return rob
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Use even & odd position:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 23.78%
func rob2(nums []int) int {
	robOdd, robEven := 0, 0

	for i := range nums {
		if i%2 == 0 {
			robEven = max(robEven+nums[i], robOdd)
		} else {
			robOdd = max(robOdd+nums[i], robEven)
		}
	}

	return max(robEven, robOdd)
}

// Use recursive with memorize:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.1 MB, less than 6.02%
func rob3(nums []int) int {
	size := len(nums)
	mem := make([]int, size)
	memView := make([]bool, size)
	return subRob(nums, mem, memView, size-1)
}

func subRob(nums, mem []int, memView []bool, index int) int {
	if index < 0 {
		return 0
	}
	if memView[index] == true {
		return mem[index]
	}

	mem[index] = max(subRob(nums, mem, memView, index-2)+nums[index], subRob(nums, mem, memView, index-1))

	memView[index] = true
	return mem[index]
}
